signature/src/balrog.rs

```diff
@@ -0,0 +1,382 @@
+    fn verify_cert_chain_pair(
+        subject: &X509Certificate,
+        issuer: &X509Certificate,
+        current_time: i64,
+    ) -> Result<(), BalrogError> {
+        /* Check that the subject and issuer names match. */
+        if subject.issuer() != issuer.subject() {
+            return Err(BalrogError::ChainSubjectMismatch);
+        }
+
+        /* Check the certificate validity period. */
+        let asn_timestamp = ASN1Time::from_timestamp(current_time).unwrap();
+        if !subject.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+        if !issuer.validity().is_valid_at(asn_timestamp) {
+            return Err(BalrogError::CertificateExpired);
+        }
+
+        /* Check that the issuer is permitted to sign certificates. */
+        let ca_key_usage = match issuer.key_usage() {
+            Err(e) => return Err(BalrogError::from(e)),
+            Ok(x) => x
+        };
+        let ca_key_usage = ca_key_usage.unwrap().value;
+        if !ca_key_usage.key_cert_sign() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+        if !issuer.is_ca() {
+            return Err(BalrogError::IssuerUnauthorized);
+        }
+
+        /* Check X509 certificate validation. */
+        let mut logger = VecLogger::default();
+        // structure validation status
+        let ok = X509StructureValidator
+            .chain(X509CertificateValidator)
+            .validate(subject, &mut logger);
+        if !ok {
+            let reason = match logger.errors().first() {
+                None => "Unknown error",
+                Some(s) => s
+            };
+            return Err(BalrogError::X509ValidationError(reason.to_string()));
+        }
+        
+        /* Verify the cryptographic signature. */
+        subject.verify_signature(Some(issuer.public_key()))
+            .map_err(|e| { BalrogError::from(e) })
+    }
+
+    /* Check a hostname against the leaf certificate. */
+    pub fn verify_leaf_hostname(&self, hostname: &str) -> Result<(), BalrogError> {
+        assert!(!self.chain.is_empty());
+        let leaf = self.chain.first().unwrap();
+
+        /* First check, does the hostname match one of the subject common names? */
+        for cn in leaf.subject().iter_common_name() {
+            let name = match cn.as_str() {
+                Err(e) => return Err(BalrogError::from(e)),
+                Ok(x) => x
+            };
+            if name == hostname {
+                return Ok(());
+            }
+        }
+
+        /* Second check, does the hostname match one of the subject alternative names? */
+        let san = match leaf.subject_alternative_name() {
+            Err(e) => return Err(BalrogError::from(e)),
+            Ok(x) => x
+        };
+        if san.is_some() {
+            let as_dns_name = GeneralName::DNSName(hostname);
+            for x in san.unwrap().value.general_names.iter() {
```

Comment:
The `unwrap` will panic if it's unsuccesfull. Let's propagate the error instead.

Reply:
This applies to all subsequent `unwrap`s.

Reply:
The unsuccessful case should be impossible here, as this path is only executed if `is_some()` returns true. But I will do some refactoring to make the naked `unwrap()`s go away.